File: parcial.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define MAX_PROCESSES 10
#define LEVELS 3  

typedef struct {
    int id;
    int burst_time;
    int priority;
    int age; 
    clock_t start_time;  
    double total_cpu_time; 
} Process;

typedef struct {
    Process *queue[MAX_PROCESSES];
    int front, rear;
} Queue;

void initializeQueue(Queue *q) {
    q->front = q->rear = -1;
}

int isEmpty(Queue *q) {
    return (q->front == -1);
}

void enqueue(Queue *q, Process *process) {
    if (q->rear == MAX_PROCESSES - 1) {
        printf("La cola está llena\n");
        exit(EXIT_FAILURE);  
    }

    if (q->front == -1)
        q->front = 0;
    q->rear++;
    q->queue[q->rear] = process;
}

Process *dequeue(Queue *q) {
    if (isEmpty(q)) {
        printf("La cola está vacía\n");
        exit(EXIT_FAILURE);
    }

    Process *process = q->queue[q->front];
    if (q->front == q->rear)
        q->front = q->rear = -1;
    else
        q->front++;
    return process;
}

void ageProcesses(Queue *queues[], int levels) {
    for (int i = 0; i < levels; ++i) {
        if (!isEmpty(queues[i])) { 
            for (int j = queues[i]->front; j <= queues[i]->rear; ++j) {
                queues[i]->queue[j]->age++;
                if (queues[i]->queue[j]->age > 5 && i > 0) {
                    queues[i]->queue[j]->priority--;
                    Process *temp = queues[i]->queue[j];
                    enqueue(queues[i-1], temp);
                    
                    for (int k = j; k < queues[i]->rear; k++) {
                        queues[i]->queue[k] = queues[i]->queue[k+1];
                    }
                    queues[i]->rear--;
                    j--; 
                }
            }
        }
    }
}

int main() {
    Queue *queues[LEVELS];  
    for (int i = 0; i < LEVELS; i++) {
        queues[i] = malloc(sizeof(Queue));
        initializeQueue(queues[i]);
    }

    Process *process1 = malloc(sizeof(Process));
    process1->id = 1;
    process1->burst_time = 5;
    process1->priority = 0;
    process1->age = 0;
    process1->total_cpu_time = 0;

    Process *process2 = malloc(sizeof(Process));
    process2->id = 2;
    process2->burst_time = 3;
    process2->priority = 1;
    process2->age = 0;
    process2->total_cpu_time = 0;

    Process *process3 = malloc(sizeof(Process));
    process3->id = 3;
    process3->burst_time = 7;
    process3->priority = 2;
    process3->age = 0;
    process3->total_cpu_time = 0;

    enqueue(queues[process1->priority], process1);
    enqueue(queues[process2->priority], process2);
    enqueue(queues[process3->priority], process3);

    printf("Ejecutando procesos en MLFQ:\n");
    while (!isEmpty(queues[0]) || !isEmpty(queues[1]) || !isEmpty(queues[2])) {
        for (int i = 0; i < LEVELS; i++) {
            while (!isEmpty(queues[i])) {
                Process *current_process = dequeue(queues[i]);
                if (current_process == NULL) continue;  
                current_process->start_time = clock();  
                printf("Proceso %d en ejecución desde la cola %d\n", current_process->id, i);
                current_process->burst_time--;
                clock_t end_time = clock(); 
                current_process->total_cpu_time += (double)(end_time - current_process->start_time) / CLOCKS_PER_SEC;
                if (current_process->burst_time > 0) {
                    enqueue(queues[i], current_process);
                } else {
                    printf("Proceso %d completado. Tiempo total de CPU: %.6f segundos.\n", current_process->id, current_process->total_cpu_time);
                    free(current_process);
                }
            }
        }
        ageProcesses(queues, LEVELS);
    }

    for (int i = 0; i < LEVELS; i++) {  // Liberar memoria de las colas
        free(queues[i]);
    }

    return 0;
}

File: parcial.py
import time

MAX_PROCESSES = 10
TIME_QUANTUM = 2
WAIT_THRESHOLD = 5

class Process:
    def __init__(self, _id, burst_time, priority):
        self.id = _id
        self.burst_time = burst_time
        self.priority = priority
        self.wait_time = 0

class Queue:
    def __init__(self):
        self.queue = []
    
    def is_empty(self):
        return len(self.queue) == 0
    
    def enqueue(self, process):
        self.queue.append(process)
    
    def dequeue(self):
        if not self.is_empty():
            return self.queue.pop(0)
        else:
            print("La cola está vacía")
            exit()

def execute_process(process, execution_time):
    print(f"Proceso {process.id} en ejecución durante {time} segundos")
    time.sleep(execution_time)

def adjust_priority(process):
    if process.wait_time >= WAIT_THRESHOLD:
        process.priority -= 1
        process.wait_time = 0
        print(f"Se ha aumentado la prioridad del proceso {process.id}")

def main():
    high_priority_queue = Queue()
    low_priority_queue = Queue()

    process1 = Process(1, 5, 1)
    process2 = Process(2, 3, 2)
    process3 = Process(3, 7, 1)

    high_priority_queue.enqueue(process1)
    low_priority_queue.enqueue(process2)
    high_priority_queue.enqueue(process3)

    print("Ejecutando procesos:")
    while not (high_priority_queue.is_empty() and low_priority_queue.is_empty()):
        if not high_priority_queue.is_empty():
            current_process = high_priority_queue.dequeue()
            execute_process(current_process, TIME_QUANTUM)
            current_process.burst_time -= TIME_QUANTUM
            current_process.wait_time += TIME_QUANTUM
            adjust_priority(current_process)
            if current_process.burst_time > 0:
                high_priority_queue.enqueue(current_process)
            else:
                del current_process

        if not low_priority_queue.is_empty():
            current_process = low_priority_queue.dequeue()
            execute_process(current_process, TIME_QUANTUM)
            current_process.burst_time -= TIME_QUANTUM
            current_process.wait_time += TIME_QUANTUM
            adjust_priority(current_process)
            if current_process.burst_time > 0:
                low_priority_queue.enqueue(current_process)
            else:
                del current_process

if __name__ == "__main__":
    main()
